### PR TITLE
Closes #325: Bucket hasher for affiliation and organization

### LINKED ISSUE #325: Bucket hasher for affiliation and organization
Create a hasher for affiliations and organizations through which they will be split into buckets in which they'll be compared with one another (cartesian).
Also, redesign the code: 
- make the AffMatchAffiliation and AffMatchOrganization be normal classes (not avro)
- create a separate AffMatchJobService and move the business logic to it.

### PR DESCRIPTION
**Pull request by lukdumi**
Info for the reviewer:

I'm going to write IisAffiliationReaderTest and IisAffiliationWriterTest in:
https://github.com/CeON/spark-utils/issues/3
After rewriting SparkAvroLoader into a normal (not util) class.

I'll write the test for AffMatchingService when it is finished.

### CODE REVIEW

**Comment by mpol** [comment]
Doesn't buid:
```
[ERROR] Failed to execute goal org.apache.avro:avro-maven-plugin:1.7.4:schema (default) on project iis-wf-affmatching: neither sourceDirectory: /home/mp/git/openaire/iis/iis-wf/iis-wf-affmatching/src/main/resources/eu/dnetlib/iis or testSourceDirectory: /home/mp/git/openaire/iis/iis-wf/iis-wf-affmatching/src/test/avro are directories -> [Help 1]
```

**Comment by lukdumi** [comment]
Should build now.

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/BucketHasher.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import java.io.Serializable;
+
+/**
+* @author Łukasz Dumiszewski
+*/
+
+public interface BucketHasher<T> extends Serializable {
+
+    
+    public String hash(T affiliation);
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/bucket/BucketHasher.java]: Parameter name? Can be affiliation or organization.

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchAffiliationNormalizer.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    private StringNormalizer countryCodeNormalizer = new LenientComparisonStringNormalizer();
+    
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    /**
+     * Normalizes properties of <code>affMatchAffiliation</code> with corresponding normalizers (see
+     * setters)<br/><br/>
+     * This method <b>changes the passed object and returns it</b>. It does NOT create a new one.
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchAffiliationNormalizer.java]: Is it safe? Spark creators discouage it: http://apache-spark-user-list.1001560.n3.nabble.com/is-it-okay-to-reuse-objects-across-RDD-s-td4877.html

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchOrganizationNormalizer.java [hunk 3]
```diff
@@ -29,44 +30,34 @@
...
     
     
     //------------------------ LOGIC --------------------------
     
-    
     /**
-     * Converts {@link Organization} into {@link AffMatchOrganization}
+     * Normalizes properties of <code>affMatchOrganization</code> with corresponding normalizers (see
+     * setters).<br/><br/>
+     * This method <b>changes the passed object and returns it</b>. It does NOT create a new one.
```
Comment (mpol) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchOrganizationNormalizer.java]: The same question as for the affiliation normalizer. Is it safe to modify RDD elements?

File: iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchAffiliationNormalizer.java (hunk 2 above)
Comment (lukdumi) [inline iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/normalize/AffMatchAffiliationNormalizer.java]: Well, in this case it was safe. It was supposed to work faster than the version that creates new objects. 
However, your doubt has convinced me that the solution with creating new objects will be cleaner so I've changed the code appropriately (we can always return to the original solution if the problems with efficiency really emerge).

**Pull request closed without merging**